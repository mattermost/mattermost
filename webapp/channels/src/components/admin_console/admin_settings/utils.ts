// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {AdminConfig, EnvironmentConfig} from '@mattermost/types/config';

function getConfigValue(config: AdminConfig | Partial<EnvironmentConfig>, path: string) {
    const pathParts = path.split('.');

    return pathParts.reduce((obj: object | null, pathPart) => {
        if (!obj) {
            return null;
        }
        return obj[(pathPart as keyof object)];
    }, config);
}

export function isSetByEnv(environmentConfig: Partial<EnvironmentConfig>, path: string) {
    return Boolean(getConfigValue(environmentConfig, path));
}

export const parseIntNonNegative = (str: string | number, defaultValue?: number) => {
    const n = typeof str === 'string' ? parseInt(str, 10) : str;

    if (isNaN(n) || n < 0) {
        if (defaultValue) {
            return defaultValue;
        }
        return 0;
    }

    return n;
};

export const parseIntZeroOrMin = (str: string | number, minimumValue = 1) => {
    const n = typeof str === 'string' ? parseInt(str, 10) : str;

    if (isNaN(n) || n < 0) {
        return 0;
    }
    if (n > 0 && n < minimumValue) {
        return minimumValue;
    }

    return n;
};

export const parseIntNonZero = (str: string | number, defaultValue?: number, minimumValue = 1) => {
    const n = typeof str === 'string' ? parseInt(str, 10) : str;

    if (isNaN(n) || n < minimumValue) {
        if (defaultValue) {
            return defaultValue;
        }
        return 1;
    }

    return n;
};
