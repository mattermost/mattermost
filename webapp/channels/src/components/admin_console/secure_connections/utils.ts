// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {LocationDescriptor} from 'history';
import {DateTime, Interval} from 'luxon';
import {useCallback, useEffect, useMemo, useState} from 'react';
import {useDispatch, useSelector} from 'react-redux';

import type {ClientError} from '@mattermost/client';
import type {Channel} from '@mattermost/types/channels';
import {isRemoteClusterPatch, type RemoteCluster, type RemoteClusterPatch} from '@mattermost/types/remote_clusters';
import type {SharedChannelRemote} from '@mattermost/types/shared_channels';
import type {Team} from '@mattermost/types/teams';
import type {IDMappedObjects, RelationOneToOne} from '@mattermost/types/utilities';

import {ChannelTypes} from 'mattermost-redux/action_types';
import {getChannel as fetchChannel} from 'mattermost-redux/actions/channels';
import {Client4} from 'mattermost-redux/client';
import {getChannel} from 'mattermost-redux/selectors/entities/channels';
import {getActiveTeamsList, getTeam} from 'mattermost-redux/selectors/entities/teams';

import type {ActionFuncAsync} from 'types/store';

export const useRemoteClusters = () => {
    const [remoteClusters, setRemoteClusters] = useState<RemoteCluster[]>();
    const [loadingState, setLoadingState] = useState<boolean | ClientError>(true);
    const {loading, error} = loadingStatus(loadingState);

    const fetch = async () => {
        setLoadingState(true);
        try {
            const data = await Client4.getRemoteClusters({excludePlugins: true});
            setRemoteClusters(data?.length ? data : undefined);
            setLoadingState(false);
        } catch (err) {
            setLoadingState(err);
        }
    };

    useEffect(() => {
        fetch();
    }, []);

    return [remoteClusters, {loading, fetch, error}] as const;
};

export const useRemoteClusterEdit = (remoteId: string | 'create', initRemoteCluster: RemoteCluster | undefined) => {
    const editing = remoteId !== 'create';

    const [currentRemoteCluster, setCurrentRemoteCluster] = useState<RemoteCluster | undefined>(initRemoteCluster);
    const [patch, setPatch] = useState<Partial<RemoteClusterPatch>>({});
    const [loading, setLoading] = useState<TLoadingState>(editing && !currentRemoteCluster);
    const [saving, setSaving] = useState<TLoadingState>(false);

    const hasChanges = Object.keys(patch).length > 0;

    useEffect(() => {
        if (!editing) {
            return;
        }
        (async () => {
            try {
                const data = await Client4.getRemoteCluster(remoteId);
                setCurrentRemoteCluster(data);
                setLoading(false);
                setPatch({});
            } catch (err) {
                setSaving(err);
            }
        })();
    }, [remoteId]);

    const applyPatch = (patch: Partial<RemoteClusterPatch>) => {
        setPatch((current) => ({...current, ...patch}));
    };

    const save = async () => {
        if (currentRemoteCluster && isRemoteClusterPatch(patch)) {
            setSaving(true);
            try {
                const data = await Client4.patchRemoteCluster(remoteId, patch);
                setCurrentRemoteCluster(data);
                setSaving(false);
                setPatch({});
            } catch (err) {
                setSaving(err);
            }
            setSaving(false);
        }
    };

    return [
        {...currentRemoteCluster, ...patch},
        {
            applyPatch,
            save,
            hasChanges,
            loading,
            saving,
            currentRemoteCluster,
            patch,
        },
    ] as const;
};

export const useSharedChannelRemotes = (remoteId: string) => {
    const [remotes, setRemotes] = useState<RelationOneToOne<Channel, SharedChannelRemote>>();
    const [loadingState, setLoadingState] = useState<TLoadingState>(true);

    const loading = isPendingState(loadingState);
    const error = !loading && loadingState;

    const fetch = async () => {
        setLoadingState(true);
        try {
            const data = await Client4.getSharedChannelRemotes(remoteId, {include_deleted: true, include_unconfirmed: true});

            setRemotes(data?.reduce<typeof remotes>((state, remote) => {
                state![remote.channel_id] = remote;

                return state;
            }, {}));
            setLoadingState(false);
        } catch (error) {
            setLoadingState(error);
        }
    };

    useEffect(() => {
        fetch();
    }, [remoteId]);

    return [remotes, {loading, error, fetch}] as const;
};

const remoteRow = (remote: SharedChannelRemote, channel: Channel, team?: Team) => {
    return {...remote, display_name: channel.display_name, team_display_name: team?.display_name ?? ''};
};

export type SharedChannelRemoteRow = SharedChannelRemote & Pick<Channel, 'display_name'> & {team_display_name: string};
export const useSharedChannelRemoteRows = (remoteId: string, opts: {filter: 'home' | 'remote' | undefined}) => {
    const [sharedChannelRemotes, setSharedChannelRemotes] = useState<SharedChannelRemoteRow[]>();
    const [loadingState, setLoadingState] = useState<TLoadingState>(true);
    const dispatch = useDispatch();

    const loading = isPendingState(loadingState);
    const error = !loading && loadingState;

    const fetch = useCallback(async () => {
        if (opts.filter === undefined) {
            // wait for a filter
            return;
        }

        setLoadingState(true);
        dispatch<ActionFuncAsync<IDMappedObjects<SharedChannelRemoteRow>>>(async (dispatch, getState) => {
            const collected: IDMappedObjects<SharedChannelRemoteRow> = {};
            const missing: SharedChannelRemote[] = [];

            try {
                const data = await Client4.getSharedChannelRemotes(remoteId, {include_unconfirmed: true, exclude_remote: opts.filter === 'home', exclude_home: opts.filter === 'remote'});
                let state = getState();
                let getMyChannelsOnce: undefined | ((remote: SharedChannelRemote) => Promise<Channel | undefined>) = async (firstNotFound: SharedChannelRemote) => {
                    const channels = await Client4.getAllTeamsChannels();
                    dispatch({
                        type: ChannelTypes.RECEIVED_ALL_CHANNELS,
                        data: channels,
                    });
                    state = getState();

                    getMyChannelsOnce = undefined; // once-only

                    // return triggering not-found remote
                    return getChannel(state, firstNotFound.channel_id);
                };

                // first-past, get known channels or do initial load on first not-found channel
                for (const remote of data) {
                    // eslint-disable-next-line no-await-in-loop
                    const channel = getChannel(state, remote.channel_id) ?? await getMyChannelsOnce?.(remote);

                    if (!channel) {
                        // collect all remotes with missing channels
                        missing?.push(remote);
                        continue;
                    }

                    const team = getTeam(state, channel.team_id);
                    collected[remote.id] = remoteRow(remote, channel, team);
                }

                // fetch missing channels individually
                if (missing.length) {
                    // TODO: performance; consider adding sharedchannelremotes search param to api/v4/channels
                    await Promise.allSettled(missing.map((remote) => dispatch(fetchChannel(remote.channel_id))));
                    state = getState();

                    for (const remote of missing) {
                        const channel = getChannel(state, remote.channel_id);

                        if (!channel) {
                            continue;
                        }

                        const team = getTeam(state, channel.team_id);
                        collected[remote.id] = remoteRow(remote, channel, team);
                    }
                }

                const rows = Object.values(collected);

                setSharedChannelRemotes(rows.length ? rows : undefined);
                setLoadingState(false);
            } catch (error) {
                setLoadingState(error);
                return {error};
            }

            return {data: collected};
        });
    }, [remoteId, opts.filter]);

    useEffect(() => {
        fetch();
    }, [remoteId, opts.filter]);

    return [sharedChannelRemotes, {loading, error, fetch}] as const;
};

export const useTeamOptions = () => {
    const teams = useSelector(getActiveTeamsList);
    const teamsById = useMemo(() => teams.reduce<IDMappedObjects<Team>>((teams, team) => ({...teams, [team.id]: team}), {}), [teams]);
    return teamsById;
};

export const getEditLocation = (rc: RemoteCluster): LocationDescriptor<RemoteCluster> => {
    return {pathname: `/admin_console/site_config/secure_connections/${rc.remote_id}`, state: rc};
};

export const getCreateLocation = (): LocationDescriptor<RemoteCluster> => {
    return {pathname: '/admin_console/site_config/secure_connections/create'};
};

const SiteURLPendingPrefix = 'pending_';
export const isConfirmed = (rc: RemoteCluster) => Boolean(rc.site_url && !rc.site_url.startsWith(SiteURLPendingPrefix));
export const isConnected = (rc: RemoteCluster) => Interval.before(DateTime.now(), {minutes: 5}).contains(DateTime.fromMillis(rc.last_ping_at));

export type TLoadingState<TError extends Error = ClientError> = boolean | TError;
export const isPendingState = <T extends Error>(loadingState: TLoadingState<T>) => loadingState === true;
export const isErrorState = <T extends Error>(loadingState: TLoadingState<T>): loadingState is T => loadingState instanceof Error;

const loadingStatus = <T extends Error>(loadingState: TLoadingState<T>) => {
    const loading = isPendingState(loadingState);
    const error = isErrorState(loadingState) ? loadingState : undefined;

    return {error, loading};
};
