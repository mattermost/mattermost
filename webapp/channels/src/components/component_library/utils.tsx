// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';

function buildPropsLists(inputPossibilities: {[x: string]: any[]}): Array<{[x: string]: any}> {
    const keys = Object.keys(inputPossibilities);
    if (!keys.length) {
        return [{}];
    }

    const selectedKey = keys[0];
    const restPossibilities = {...inputPossibilities};
    delete restPossibilities[selectedKey];
    const subProps = buildPropsLists(restPossibilities);
    const result: Array<{[x: string]: any}> = [];
    inputPossibilities[selectedKey].forEach((v) => {
        subProps.forEach((rest) => {
            result.push({...rest, [selectedKey]: v});
        });
    });

    return result;
}

function buildPropString(inputProps: {[x: string]: any}) {
    const propKeys = Object.keys(inputProps);
    if (!propKeys.length) {
        return undefined;
    }

    const result = [(<>{'PROPS: '}</>)];
    propKeys.forEach((v) => {
        result.push((<><b>{v}</b>{`: ${inputProps[v]}, `}</>));
    });
    return result;
}

export function buildComponent(
    Component: React.ComponentType<any>,
    propPossibilities: {[x: string]: any[]},
    dropdownPossibilities: Array<{[x: string]: string[]} | undefined>,
    setProps: Array<{[x: string]: any} | undefined>,
) {
    const res: React.ReactNode[] = [];
    let currentPropPossibilities = {...propPossibilities};
    dropdownPossibilities.forEach((v) => {
        if (v) {
            currentPropPossibilities = {
                ...currentPropPossibilities,
                ...v,
            };
        }
    });

    const propsVariations = buildPropsLists(currentPropPossibilities);
    let builtSetProps = {};
    setProps.forEach((v) => {
        if (v) {
            builtSetProps = {
                ...builtSetProps,
                ...v,
            };
        }
    });
    propsVariations.forEach((v) => {
        const propString = buildPropString(v);
        res.push(
            <>
                {Boolean(propString) && <p>{propString}</p>}
                <Component
                    {...builtSetProps}
                    {...v}
                />
            </>,
        );
    });
    return res;
}
