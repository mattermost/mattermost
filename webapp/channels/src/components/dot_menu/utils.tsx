// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type React from 'react';

export type ChangeEvent = React.KeyboardEvent | React.MouseEvent;

export const trackDotMenuEvent = (): void => {
    // Telemetry tracking removed
};
