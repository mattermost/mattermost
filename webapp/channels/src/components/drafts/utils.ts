// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {IntlShape} from 'react-intl';
import {defineMessages} from 'react-intl';

import type {ScheduledPostErrorCode} from '@mattermost/types/schedule_post';

const errorCodeToErrorMessage = defineMessages<ScheduledPostErrorCode>({
    unknown: {
        id: 'scheduled_post.error_code.unknown_error',
        defaultMessage: 'Unknown Error',
    },
    channel_archived: {
        id: 'scheduled_post.error_code.channel_archived',
        defaultMessage: 'Channel Archived',
    },
    channel_not_found: {
        id: 'scheduled_post.error_code.channel_removed',
        defaultMessage: 'Channel Removed',
    },
    user_missing: {
        id: 'scheduled_post.error_code.user_missing',
        defaultMessage: 'User Deleted',
    },
    user_deleted: {
        id: 'scheduled_post.error_code.user_deleted',
        defaultMessage: 'User Deleted',
    },
    no_channel_permission: {
        id: 'scheduled_post.error_code.no_channel_permission',
        defaultMessage: 'Missing Permission',
    },
    no_channel_member: {
        id: 'scheduled_post.error_code.no_channel_member',
        defaultMessage: 'Not In Channel',
    },
    thread_deleted: {
        id: 'scheduled_post.error_code.thread_deleted',
        defaultMessage: 'Thread Deleted',
    },
    unable_to_send: {
        id: 'scheduled_post.error_code.unable_to_send',
        defaultMessage: 'Unable to Send',
    },
    invalid_post: {
        id: 'scheduled_post.error_code.invalid_post',
        defaultMessage: 'Invalid Post',
    },
});

export function getErrorStringFromCode(intl: IntlShape, errorCode: ScheduledPostErrorCode = 'unknown') {
    const textDefinition = errorCodeToErrorMessage[errorCode] ?? errorCodeToErrorMessage.unknown;
    return intl.formatMessage(textDefinition).toUpperCase();
}
