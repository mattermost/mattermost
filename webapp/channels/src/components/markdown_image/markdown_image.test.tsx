// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {shallow} from 'enzyme';
import React from 'react';

import FilePreviewModal from 'components/file_preview_modal';
import SizeAwareImage from 'components/size_aware_image';

import Constants from 'utils/constants';
import {TestHelper} from 'utils/test_helper';

import MarkdownImage from './markdown_image';

describe('components/MarkdownImage', () => {
    const baseProps = {
        imageMetadata: {
            format: 'png',
            height: 90,
            width: 1041,
            frameCount: 0,
        },
        alt: 'test image',
        height: '',
        width: '',
        title: 'test title',
        className: 'markdown-inline-img',
        postId: 'post_id',
        imageIsLink: false,
        onImageLoaded: jest.fn(),
        onImageHeightChanged: jest.fn(),
        postType: 'system_generic',
        actions: {
            openModal: jest.fn(),
        },
    };

    test('should match snapshot', () => {
        const props = {...baseProps, src: '/images/logo.png'};
        const wrapper = shallow(
            <MarkdownImage {...props}/>,
        );

        expect(wrapper).toMatchSnapshot();
    });

    test('should match snapshot for broken link', () => {
        const props = {...baseProps,
            imageMetadata: {
                format: 'png',
                height: 10,
                width: 10,
                frameCount: 0,
            },
            src: 'brokenLink'};
        const wrapper = shallow(
            <MarkdownImage {...props}/>,
        );

        expect(wrapper).toMatchSnapshot();
    });

    test('should handle load failure properly', () => {
        const props = {...baseProps,
            imageMetadata: {
                format: 'png',
                height: 10,
                width: 10,
                frameCount: 0,

            },
            src: 'brokenLink'};
        const wrapper = shallow(
            <MarkdownImage {...props}/>,
        );

        expect(wrapper.state('loadFailed')).toBe(false);

        (wrapper.instance() as MarkdownImage).handleLoadFail();

        expect(wrapper.state('loadFailed')).toBe(true);
    });

    test('should reset loadFailed state after image source is updated', () => {
        const props = {...baseProps,
            imageMetadata: {
                format: 'png',
                height: 10,
                width: 10,
                frameCount: 0,

            },
            src: 'brokenLink'};
        const nextProps = {...baseProps, src: 'https://example.com/image.png'};
        const wrapper = shallow(
            <MarkdownImage {...props}/>,
        );

        wrapper.instance().setState({loadFailed: true});

        wrapper.setProps(nextProps);

        expect(wrapper.state('loadFailed')).toBe(false);
    });

    test('should render a link if the source is unsafe', () => {
        const props = {...baseProps, src: ''};
        const wrapper = shallow(
            <MarkdownImage {...props}/>,
        );
        expect(wrapper.find('img').props().alt).toBe(props.alt);
        expect(wrapper.find('img').props().className).toBe(props.className + ' broken-image');
        expect(wrapper).toMatchSnapshot();
    });

    test('should handle not loaded state properly', () => {
        const props = {...baseProps, src: 'https://example.com/image.png'};
        const wrapper = shallow(
            <MarkdownImage {...props}/>,
        );

        expect(wrapper.state('loaded')).toBe(false);

        const childrenNode = wrapper.props().children(props.src);

        // using a div as a workaround because shallow doesn't support react fragments
        const childrenWrapper = shallow(<div>{childrenNode}</div>);

        expect(childrenWrapper.find(SizeAwareImage)).toHaveLength(1);
        expect(childrenWrapper.find(SizeAwareImage).prop('className')).
            toEqual(`${props.className} markdown-inline-img--loading`);
    });

    test('should handle not loaded state properly in case of a header change system message', () => {
        const props = {...baseProps, src: 'https://example.com/image.png', postType: 'system_header_change'};
        const wrapper = shallow(
            <MarkdownImage {...props}/>,
        );

        expect(wrapper.state('loaded')).toBe(false);

        const childrenNode = wrapper.props().children(props.src);

        // using a div as a workaround because shallow doesn't support react fragments
        const childrenWrapper = shallow(<div>{childrenNode}</div>);

        expect(childrenWrapper.find(SizeAwareImage)).toHaveLength(1);
        expect(childrenWrapper.find(SizeAwareImage).prop('className')).
            toEqual(`${props.className} markdown-inline-img--scaled-down-loading`);
    });

    test('should set loaded state when img loads and call onImageLoaded prop', () => {
        const props = {...baseProps, src: 'https://example.com/image.png'};
        const wrapper = shallow(
            <MarkdownImage {...props}/>,
        );
        const dimensions = {
            height: props.imageMetadata.height,
            width: props.imageMetadata.width,
        };

        expect(wrapper.state('loaded')).toBe(false);

        (wrapper.instance() as MarkdownImage).handleImageLoaded(dimensions);

        expect(wrapper.state('loaded')).toBe(true);

        expect(props.onImageLoaded).toHaveBeenCalledTimes(1);
        expect(props.onImageLoaded).toHaveBeenCalledWith(dimensions);
    });

    it('should match snapshot for SizeAwareImage dimensions', () => {
        const props = {...baseProps,
            imageMetadata: {format: 'jpg', frameCount: 0, width: 100, height: 90},
            src: 'path/image',
        };
        const wrapper = shallow(
            <MarkdownImage {...props}/>,
        );

        wrapper.instance().setState({loaded: true});

        const childrenWrapper = wrapper.props().children('safeSrc');
        expect(childrenWrapper).toMatchSnapshot();
    });

    test('should render an image with preview modal if the source is safe', () => {
        const props = {...baseProps, src: 'https://example.com/image.png'};
        const wrapper = shallow(
            <MarkdownImage {...props}/>,
        );
        wrapper.instance().setState({loaded: true});

        const childrenNode = wrapper.props().children(props.src);

        // using a div as a workaround because shallow doesn't support react fragments
        const childrenWrapper = shallow(<div>{childrenNode}</div>);

        expect(childrenWrapper.find(SizeAwareImage)).toHaveLength(1);
        expect(childrenWrapper.find(SizeAwareImage).prop('className')).
            toEqual(`${props.className} markdown-inline-img--hover cursor--pointer a11y--active`);
        expect(childrenWrapper).toMatchSnapshot();
    });

    test('should render an image with no preview if the source is safe and the image is a link', () => {
        const props = {...baseProps, src: 'https://example.com/image.png', imageIsLink: true};
        const wrapper = shallow(
            <MarkdownImage {...props}/>,
        );
        wrapper.instance().setState({loaded: true});

        const childrenNode = wrapper.props().children(props.src);

        // using a div as a workaround because shallow doesn't support react fragments
        const childrenWrapper = shallow(<div>{childrenNode}</div>);

        expect(childrenWrapper.find(SizeAwareImage)).toHaveLength(1);
        expect(childrenWrapper.find(SizeAwareImage).prop('className')).
            toEqual(`${props.className} markdown-inline-img--hover markdown-inline-img--no-border`);
        expect(childrenWrapper.find(FilePreviewModal)).toHaveLength(0);
        expect(childrenWrapper).toMatchSnapshot();
    });

    test('should call openModal when showModal is called', () => {
        const props = {...baseProps, src: 'https://example.com/image.png'};
        const wrapper = shallow(
            <MarkdownImage {...props}/>,
        );
        const mockEvent = {
            preventDefault: () => {},
        };
        (wrapper.instance() as MarkdownImage).showModal(mockEvent as unknown as React.MouseEvent<HTMLImageElement>, 'https://example.com/image.png');
        expect(props.actions.openModal).toHaveBeenCalledTimes(1);
    });

    test('should use provided fileInfo when opening modal', () => {
        const mockFileInfo = TestHelper.getFileInfoMock({
            id: 'file123',
            name: 'test.gif',
            extension: 'gif',
            size: 1024,
            has_preview_image: true,
            link: 'https://example.com/image.png',
        });
        const props = {...baseProps, src: 'https://example.com/image.png', fileInfo: mockFileInfo};
        const wrapper = shallow(
            <MarkdownImage {...props}/>,
        );
        const mockEvent = {
            preventDefault: () => {},
        };
        (wrapper.instance() as MarkdownImage).showModal(mockEvent as unknown as React.MouseEvent<HTMLImageElement>, 'https://example.com/image.png');

        expect(props.actions.openModal).toHaveBeenCalledWith({
            modalId: 'file_preview_modal',
            dialogType: FilePreviewModal,
            dialogProps: {
                startIndex: 0,
                postId: props.postId,
                fileInfos: [mockFileInfo],
            },
        });
    });

    test('should create minimal fileInfo when none provided', () => {
        const props = {...baseProps, src: 'https://example.com/image.png'};
        const wrapper = shallow(
            <MarkdownImage {...props}/>,
        );
        const mockEvent = {
            preventDefault: () => {},
        };
        (wrapper.instance() as MarkdownImage).showModal(mockEvent as unknown as React.MouseEvent<HTMLImageElement>, 'https://example.com/image.png');

        expect(props.actions.openModal).toHaveBeenCalledWith({
            modalId: 'file_preview_modal',
            dialogType: FilePreviewModal,
            dialogProps: {
                startIndex: 0,
                postId: props.postId,
                fileInfos: [{
                    has_preview_image: false,
                    link: 'https://example.com/image.png',
                    extension: 'png',
                    name: props.alt,
                }],
            },
        });
    });

    test('should properly scale down the image in case of a header change system message', () => {
        const props = {...baseProps, src: 'https://example.com/image.png', postType: 'system_header_change'};
        const wrapper = shallow(
            <MarkdownImage {...props}/>,
        );
        wrapper.instance().setState({loaded: true});

        const childrenNode = wrapper.props().children(props.src);

        // using a div as a workaround because shallow doesn't support react fragments
        const childrenWrapper = shallow(<div>{childrenNode}</div>);

        expect(childrenWrapper.find(SizeAwareImage)).toHaveLength(1);
        expect(childrenWrapper.find(SizeAwareImage).prop('className')).
            toEqual(`${props.className} markdown-inline-img--hover cursor--pointer a11y--active markdown-inline-img--scaled-down`);
    });

    test('should render image with title, height, width', () => {
        const props = {
            alt: 'test image',
            title: 'test title',
            className: 'markdown-inline-img',
            postId: 'post_id',
            src: 'https://example.com/image.png',
            imageIsLink: false,
            height: '76',
            width: '50',
            actions: baseProps.actions,
        };

        const wrapper = shallow(
            <MarkdownImage {...props}/>,
        );
        wrapper.instance().setState({loaded: true});

        const childrenNode = wrapper.props().children(props.src);

        // using a div as a workaround because shallow doesn't support react fragments
        const childrenWrapper = shallow(<div>{childrenNode}</div>);

        expect(childrenWrapper.find(SizeAwareImage)).toHaveLength(1);
        expect(childrenWrapper.find(SizeAwareImage).prop('className')).
            toEqual(`${props.className} markdown-inline-img--hover cursor--pointer a11y--active`);

        // These props can be strings since they're passed directly through to HTML tags
        expect(childrenWrapper.find(SizeAwareImage).prop('width')).toEqual('50');
        expect(childrenWrapper.find(SizeAwareImage).prop('height')).toEqual('76');

        expect(childrenWrapper.find(SizeAwareImage).prop('title')).toEqual('test title');
    });

    test(`should render image with MarkdownImageExpand if it is taller than ${Constants.EXPANDABLE_INLINE_IMAGE_MIN_HEIGHT}px`, () => {
        const props = {
            alt: 'test image',
            title: 'test title',
            className: 'markdown-inline-img',
            postId: 'post_id',
            src: 'https://example.com/image.png',
            imageIsLink: false,
            height: '250',
            width: '50',
            actions: baseProps.actions,
        };

        const wrapper = shallow(
            <MarkdownImage {...props}/>,
        );
        wrapper.instance().setState({loaded: true});
        const childrenNode = wrapper.props().children(props.src);

        // using a div as a workaround because shallow doesn't support react fragments
        const childrenWrapper = shallow(<div>{childrenNode}</div>);

        expect(childrenWrapper).toMatchSnapshot();
    });

    test('should provide image src as an alt text for MarkdownImageExpand if image has no own alt text', () => {
        const props = {
            alt: '',
            title: 'test title',
            className: 'markdown-inline-img',
            postId: 'post_id',
            src: 'https://example.com/image.png',
            imageIsLink: false,
            height: '250',
            width: '50',
            actions: baseProps.actions,
        };

        const wrapper = shallow(
            <MarkdownImage {...props}/>,
        );
        wrapper.instance().setState({loaded: true});
        const childrenNode = wrapper.props().children(props.src);

        // using a div as a workaround because shallow doesn't support react fragments
        const childrenWrapper = shallow(<div>{childrenNode}</div>);

        expect(childrenWrapper).toMatchSnapshot();
    });

    test('should render a alt text if the link is unsafe', () => {
        const props = {...baseProps, isUnsafeLinksPost: true};
        const wrapper = shallow(
            <MarkdownImage {...props}/>,
        );
        expect(wrapper.text()).toBe(props.alt);
    });

    test('should pass fileInfo prop to SizeAwareImage when provided', () => {
        const fileInfo = TestHelper.getFileInfoMock({
            id: 'file123',
            name: 'test.gif',
            extension: 'gif',
            mime_type: 'image/gif',
        });
        const props = {
            ...baseProps,
            src: 'https://example.com/test.gif',
            fileInfo,
        };

        const wrapper = shallow(
            <MarkdownImage {...props}/>,
        );
        wrapper.instance().setState({loaded: true});

        const childrenNode = wrapper.props().children(props.src);
        const childrenWrapper = shallow(<div>{childrenNode}</div>);

        const sizeAwareImage = childrenWrapper.find(SizeAwareImage);
        expect(sizeAwareImage).toHaveLength(1);
        expect(sizeAwareImage.prop('fileInfo')).toBe(fileInfo);
    });

    test('should pass undefined fileInfo to SizeAwareImage when not provided', () => {
        const props = {
            ...baseProps,
            src: 'https://example.com/test.gif',

            // No fileInfo prop
        };

        const wrapper = shallow(
            <MarkdownImage {...props}/>,
        );
        wrapper.instance().setState({loaded: true});

        const childrenNode = wrapper.props().children(props.src);
        const childrenWrapper = shallow(<div>{childrenNode}</div>);

        const sizeAwareImage = childrenWrapper.find(SizeAwareImage);
        expect(sizeAwareImage).toHaveLength(1);
        expect(sizeAwareImage.prop('fileInfo')).toBeUndefined();
    });
});
