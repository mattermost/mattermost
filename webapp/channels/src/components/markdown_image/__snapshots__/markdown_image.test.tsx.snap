// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/MarkdownImage should match snapshot 1`] = `
<Connect(Component)
  imageMetadata={
    Object {
      "format": "png",
      "frameCount": 0,
      "height": 90,
      "width": 1041,
    }
  }
  src="/images/logo.png"
>
  <Component />
</Connect(Component)>
`;

exports[`components/MarkdownImage should match snapshot for SizeAwareImage dimensions 1`] = `
<injectIntl(SizeAwareImage)
  alt="test image"
  className="markdown-inline-img markdown-inline-img--hover markdown-inline-img--no-border"
  dimensions={
    Object {
      "format": "jpg",
      "frameCount": 0,
      "height": 90,
      "width": 100,
    }
  }
  height=""
  onClick={[Function]}
  onImageLoadFail={[Function]}
  onImageLoaded={[Function]}
  showLoader={false}
  src="safeSrc"
  title="test title"
  width=""
/>
`;

exports[`components/MarkdownImage should match snapshot for broken link 1`] = `
<Connect(Component)
  imageMetadata={
    Object {
      "format": "png",
      "frameCount": 0,
      "height": 10,
      "width": 10,
    }
  }
  src="brokenLink"
>
  <Component />
</Connect(Component)>
`;

exports[`components/MarkdownImage should provide image src as an alt text for MarkdownImageExpand if image has no own alt text 1`] = `
<div>
  <Connect(MarkdownImageExpand)
    alt="https://example.com/image.png"
    imageKey="https://example.com/image.png"
    postId="post_id"
  >
    <injectIntl(SizeAwareImage)
      alt=""
      className="markdown-inline-img markdown-inline-img--hover cursor--pointer a11y--active"
      dimensions={Object {}}
      height="250"
      onClick={[Function]}
      onImageLoadFail={[Function]}
      onImageLoaded={[Function]}
      showLoader={false}
      src="https://example.com/image.png"
      title="test title"
      width="50"
    />
  </Connect(MarkdownImageExpand)>
</div>
`;

exports[`components/MarkdownImage should render a link if the source is unsafe 1`] = `
<div
  style={
    Object {
      "display": "inline-block",
    }
  }
>
  <img
    alt="test image"
    className="markdown-inline-img broken-image"
    src=""
  />
</div>
`;

exports[`components/MarkdownImage should render an image with no preview if the source is safe and the image is a link 1`] = `
<div>
  <injectIntl(SizeAwareImage)
    alt="test image"
    className="markdown-inline-img markdown-inline-img--hover markdown-inline-img--no-border"
    dimensions={
      Object {
        "format": "png",
        "frameCount": 0,
        "height": 90,
        "width": 1041,
      }
    }
    height=""
    onClick={[Function]}
    onImageLoadFail={[Function]}
    onImageLoaded={[Function]}
    showLoader={false}
    src="https://example.com/image.png"
    title="test title"
    width=""
  />
</div>
`;

exports[`components/MarkdownImage should render an image with preview modal if the source is safe 1`] = `
<div>
  <injectIntl(SizeAwareImage)
    alt="test image"
    className="markdown-inline-img markdown-inline-img--hover cursor--pointer a11y--active"
    dimensions={
      Object {
        "format": "png",
        "frameCount": 0,
        "height": 90,
        "width": 1041,
      }
    }
    height=""
    onClick={[Function]}
    onImageLoadFail={[Function]}
    onImageLoaded={[Function]}
    showLoader={false}
    src="https://example.com/image.png"
    title="test title"
    width=""
  />
</div>
`;

exports[`components/MarkdownImage should render image with MarkdownImageExpand if it is taller than 100px 1`] = `
<div>
  <Connect(MarkdownImageExpand)
    alt="test image"
    imageKey="https://example.com/image.png"
    postId="post_id"
  >
    <injectIntl(SizeAwareImage)
      alt="test image"
      className="markdown-inline-img markdown-inline-img--hover cursor--pointer a11y--active"
      dimensions={Object {}}
      height="250"
      onClick={[Function]}
      onImageLoadFail={[Function]}
      onImageLoaded={[Function]}
      showLoader={false}
      src="https://example.com/image.png"
      title="test title"
      width="50"
    />
  </Connect(MarkdownImageExpand)>
</div>
`;
